src/shred_network/collector/shred_receiver.zig

```diff
@@ -186,13 +201,344 @@ pub const ShredReceiver = struct {
             },
         );
         self.metrics.passed_to_inserter_count.add(self.shred_batch.len);
+
+        try self.handleDuplicateSlots(allocator, &result);
+
         result.deinit();
 
         self.metrics.batch_size.observe(self.shred_batch.len);
     }
 
     const MAX_SHREDS_PER_ITER = 1024;
 
+    /// Handles detected duplicate slots by:
+    /// - Send duplicate slot notifications to be handled in consensus part of replay
+    /// - Store the duplicate proof in the ledger
+    /// - Broadcast the duplicate proof via gossip
+    /// Analogous to [check_duplicate](https://github.com/anza-xyz/agave/blob/aa2f078836434965e1a5a03af7f95c6640fe6e1e/core/src/window_service.rs#L155)
+    fn handleDuplicateSlots(
```

Comment:
this code isn't about receiving shreds, so i find it unintuitive/unexpected to find this code within the shred receiver. can you move this implementation to another file? maybe create a struct called DuplicateShredHandler that contains all the necessary dependencies, and ShredReceiver can accept it as a param just like ShredInserter.

Reply:
[991a0c6](https://github.com/Syndica/sig/pull/1027/commits/991a0c69c69cf029251cfd2bb61a658825f6db5b)